learn-data-structures-by-building-a-shopping-cart/index11b.js

```diff
@@ -0,0 +1,56 @@
+const products = [
+  {
+    id: 0,
+    name: "Vanilla buttercream cupcake",
+    price: 2.99,
+  },
+  {
+    id: 1,
+    name: "French Macaroon",
+    price: 3.99,
+  },
+  {
+    id: 2,
+    name: "Fruit sprinkles cupcake",
+    price: 3.99,
+  },
+  {
+    id: 3,
+    name: "Pink flower cupcake",
+    price: 5.99,
+  },
+];
+
+function shoppingCartGenerator() {
+  const newShoppingCart = {};
+  newShoppingCart.items = [];
+  newShoppingCart.addItem = (productId) => {
+    for (let i = 0; i < products.length; i++) {
+      if (productId === products[i].id) {
+        shoppingCart.items.push(productId);
+        break;
+      }
+    }
+  };
+  newShoppingCart.getItems = () => {
+    return shoppingCart.items;
+  };
+  return newShoppingCart;
+}
+
+const primary = shoppingCartGenerator();
+const wishlist = shoppingCartGenerator();
+
+/*
+  A smart approach would be to move the functions `getItems` and `addItem` to an object that can be used
```

Comment:
Your explanations in the last two steps were very good. Unfortunately they're a bit too long, and there are no code changes for us to test. If possible, we should shorten the explanation to something like the following:

Unfortunately we have a problem -- every time an object is created using `shoppingCartGenerator`, we are creating copies of the object's code in memory.
Inside each shopping cart, we have separate copies of the `items` property, `addItem` method and `getItems` method.
If we add more methods and create more shopping carts, memory will be used up more quickly and limit the performance of our computer/browser.
A smart approach would be to move the functions `getItems` and `addItem` to an object that can be used to store all the methods we wish our `shoppingCart` objects to have access to:
```
const functionStore = {
  method1: () => {
    console.log('Hey, I am a function')
  }
}
```
Create an empty object called `functionStore`.